### ISSUE TITLE
Serverside / Linux support

### ISSUE THREAD

**Issue by mfreiwald** [jamesrochabrun/SwiftOpenAI#89]
Any plans to make this buildable on Linux?
Since its better to not put your API-keys on the client I want to use this on my swift server project.

I created a [fork](https://github.com/mfreiwald/SwiftOpenAI/tree/serverside) and replaced `URLSession` by `HTTPClient` from [async-http-client](https://github.com/swift-server/async-http-client)
but it's not fully tested yet and I have no idea yet what is the best approach to make it available for `URLSession` AND `HTTPClient`.

**Comment by jamesrochabrun** [comment]
Not plans at the moment, open for collaborations!

**Comment by mergesort** [comment]
I was just doing a comparison of OpenAI-compatible packages to integrate into my Vapor app, and came across this project. I've already integrated it into a local tool and think you've got a great API surface, I quite prefer it the ergonomics of this library to the [MacPaw OpenAI](https://github.com/MacPaw/OpenAI) library. Is there any chance you'd reconsider adding Linux support?

I've helped add Linux networking support to [other projects](https://github.com/will-lumley/FaviconFinder) before, and it's not particularly difficult to maintain, after the initial work is done you can automate the process of making sure things still work by [building on Linux](https://github.com/will-lumley/FaviconFinder/blob/main/.github/workflows/BuildTests-linux.yml) in a Github Action.

**Comment by jamesrochabrun** [comment]
Thanks for the interest, I am trying to get the issues as minimum while trying to handle other responsibilities 🙏

But lets see how we can move forward here.

@mfreiwald is it correct to say that this is all we need? https://github.com/jamesrochabrun/SwiftOpenAI/compare/main...mfreiwald:SwiftOpenAI:serverside

Why do we need to remove [Sources/OpenAI/AIProxy/AIProxyCertificatePinning.swift](https://github.com/jamesrochabrun/SwiftOpenAI/compare/main...mfreiwald:SwiftOpenAI:serverside#diff-92eeabc9e08b1646e396141f0024b154a7d6cc7e5837a47af09369238d990421)

I will like to add more functionality but without sacrificing existing one, please let me know your thoughts, thanks.

**Comment by mfreiwald** [comment]
No, my solution replaces URLSession completely with AsyncHttpClient. 
a smaller change would be to adjust the URLSession calls so that they align with ‚FoundationNetworking‘. 

I try to find some time in the next two weeks for a better solution including a PR.

**Comment by mergesort** [comment]
My suggestion would be to not depend on the Linux implementation of URLSession, but to instead continue integrating AsyncHTTPClient. There are many bugs in the Linux implementation of `URLSession`, not just minor ones, and they will continue to persist now that URLSession is unmaintained in favor of AsyncHTTPClient.

What I would recommend doing is isolating the networking logic, and exposing a common interface that layers over URLSession and AsyncHTTPClient in a way that the rest of your callers never need to think about. This would make it so only one file needs to change a lot, but every other file in your library (and external users) wouldn't have to be concerned with these implementation details.

If that makes sense, I'm happy to talk through this some more!

**Comment by mergesort** [comment]
Hi @jamesrochabrun, I finally had some time to explore this and came up with a simple solution.

[This change](https://github.com/mergesort/SwiftOpenAI/commit/74e9ec87688df190bd2473c419adb6ee6495c8a0) looks like a monster, but for the most part it's just spacing changes, because as best I can tell your project seems to use 3 space tabs versus most 2 or 4 space configurations.

In reality all this change does is:
- Swaps in [AsyncHTTPClient](https://github.com/AsyncHttpClient/async-http-client) for URLSession. This is becoming a standard in most multi-platform projects I've seen, because URLSession is horribly broken on Linux.
- The one exception is anything related to AIProxy. I've left everything alone there as best as I can, and simply do not compile the code on Linux, since AIProxy does not make much sense to use in a server environment where API requests are already protected.

If you think this is a good path forward and would be willing to accept a PR, I would be very happy to clean this up to your liking, I really would love to use this on Linux. (I can also do the same for SwiftAnthropic if you'd be ok with it, I really would love to be able to use that as well.)

Please let me know, and thanks again for all your work on this great library!

**Comment by gsabran** [comment]
@mergesort could you create a PR? I was talking with @jamesrochabrun. Your change overall looks great, but we were hesitant to migrate everyone off URLSession that for all its issues is a reliable workhorse on non linux platforms. Do you think you could make URLSession conform to the new `HTTPClient` protocol, and default non linux users to that existing implementation, so that this change is transparent to them?



FYI I just made https://github.com/jamesrochabrun/SwiftOpenAI/pull/141 that'll make spacing consistent. You probably want to cherry-pick https://github.com/jamesrochabrun/SwiftOpenAI/pull/141/commits/603e32c3fdd8fff2bd782d7ef53cfefaa6c0d2b5 and run the linter on your end to avoid massive merge conflicts.

**Comment by mergesort** [comment]
That's a great idea, conforming URLSession to `HTTPClient` is better than the what I had in mind! I'll start from scratch with the new spacing and clean up everything, then file a PR for review.

**Comment by mergesort** [comment]
Done in #143, let me know what you think!

**Comment by mergesort** [comment]
Hey @jamesrochabrun! Yep, I've been thinking about it since I opened [#49](https://github.com/jamesrochabrun/SwiftAnthropic/issues/49), but my schedule has gotten super packed before a much-needed vacation. I'm back now and have a couple of hectic weeks of work coming up, but I hope I can get to it later this summer if that's ok.

**Comment by jamesrochabrun** [comment]
@mergesort I vibe code it https://github.com/jamesrochabrun/SwiftAnthropic/pull/50
Basically I made Claude Code apply the same changes you added in SwiftOpenAI in SwiftAnthropic, as the network layers were very similar, no rush but if you have a sec it will be awesome if you can take a look on the PR.
again no rush and thanks so much for your contribution.
